### PR TITLE
Add support for `:include` index option

### PR DESCRIPTION
**Pull request by steve-abrams**
### Summary

This adds the option `:include` to PostgreSQL btree indexes to include non-key columns in the index using the `INCLUDE` parameter.

The `INCLUDE` parameter of `CREATE INDEX` is described in the [PostgreSQL docs](https://www.postgresql.org/docs/current/sql-createindex.html):

> The optional INCLUDE clause specifies a list of columns which will be included in the index as non-key columns.

Example:

```ruby
# In a migration
def change
  add_index :users, :email, include: :id
end
```

will result in:

```sql
CREATE INDEX index_users_on_email USING btree (email) INCLUDE (id)
```

Example with multiple columns:

```ruby
# In a migration
def change
  add_index :users, :email, include: [:id, :created_at]
end
```

will result in:

```sql
CREATE INDEX index_users_on_email USING btree (email) INCLUDE (id, created_at)
```

### CODE REVIEW

**Comment by steve-abrams** [comment]
👋 Hello! This is my first time contributing to Rails. I'm a little uncertain if I missed some tests that need updating, so please do point me in the right direction if I missed anything.

File: activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb
Comment (eileencodes) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb:105]: Is it possible to handle this in the postgres schema_creation file (activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb) instead since this is a postgres only feature?

Comment (eileencodes) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb:105]: Is it possible to handle this in the postgres schema_creation file (activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb) instead since this is a postgres only feature?

File: activerecord/test/schema/schema.rb
Comment (eileencodes) [inline activerecord/test/schema/schema.rb:278]: There's a postgres specific schema in activerecord/test/schema/postgresql_specific_schema.rb which might be a more appropriate file for this change since this won't work on any other adapters.

Comment (eileencodes) [inline activerecord/test/schema/schema.rb:278]: There's a postgres specific schema in activerecord/test/schema/postgresql_specific_schema.rb which might be a more appropriate file for this change since this won't work on any other adapters.

File: activerecord/test/cases/schema_dumper_test.rb
Comment (eileencodes) [inline activerecord/test/cases/schema_dumper_test.rb:203]: I think this test should be moved to activerecord/test//cases/adapters/postgresql/schema_test.rb after moving the schema change below. This feature is postgres specific so it should be contained to postgres files as much as possible.

Comment (eileencodes) [inline activerecord/test/cases/schema_dumper_test.rb:203]: I think this test should be moved to activerecord/test//cases/adapters/postgresql/schema_test.rb after moving the schema change below. This feature is postgres specific so it should be contained to postgres files as much as possible.

**Review by eileencodes — CHANGES_REQUESTED** [review]
Congrats on your first PR @steve-abrams! I left some comments about changes we need. The tests look fine other than location, I think the ones I commented on should be moved to the postgres specific tests.

File: activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb
Comment (steve-abrams) [reply activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb]: Good point, I've removed it. I also realized, this option is supported starting in version 11 of Postgres, so I updated the postgresql version of this method to account for that.

File: activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb
Comment (steve-abrams) [reply activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb]: Yes, I've moved it and the associated methods to the postgres `schema_creation`  file. I also moved `quoted_include_columns_for_index` to the postgres `schema_statements` file.

File: activerecord/test/schema/schema.rb
Comment (steve-abrams) [reply activerecord/test/schema/schema.rb]: 👍 I moved the index to `activerecord/test/schema/postgresql_specific_schema.rb`.

File: activerecord/test/cases/schema_dumper_test.rb
Comment (steve-abrams) [reply activerecord/test/cases/schema_dumper_test.rb]: Thanks, I've moved the test.

**Comment by steve-abrams** [comment]
Thanks for the feedback @eileencodes, I've made the suggested changes and I _think_ I've moved everything I can to postgres-specific files. Let me know what you think!

**Comment by fatkodima** [comment]
There is an existing PR with this feature https://github.com/rails/rails/pull/37515 from @sebastian-palma

**Comment by eileencodes** [comment]
> There is an existing PR with this feature https://github.com/rails/rails/pull/37515 from @sebastian-palma

We can give them credit on this PR. Reopening a PR from 2020 means someone is going to have to fix the conflicts.

That aside, I need someone more familiar with Postgres to review this as I really don't have experience with this adapter/client and consequences of this feature.

**Review by eileencodes — APPROVED** [review]
This looks good to me, just need someone with more postgres experience to offer an opinion on the feature.

**Comment by steve-abrams** [comment]
> There is an existing PR with this feature https://github.com/rails/rails/pull/37515 from @sebastian-palma

>> We can give them credit on this PR.

I'm happy to add credit, would I just add @sebastian-palma to the changelog? 

----- 

I also went ahead and squashed, rebased, and fixed up the conflicts for the next reviewer.

**Review by eileencodes — COMMENTED** [review]
You would make the suggested change I left in the changelog and then add the following to your commit message:

```
Co-authored-by: Sebastián Palma <vnhnhm@gmail.com>
```

(Got Sebastián's email from Rails commit history, github has no way of finding it in the UI but coauthored doesnt work without it).

**Comment by sebastian-palma** [comment]
All yours @steve-abrams thanks for taking the time to push for this, no credits needed. Thanks @eileencodes , @fatkodima for remembering about it

File: activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb
Comment (fatkodima) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:810]: `UNIQUE` is extra.

> with btree indexes

That was in 11 version, now also `GIST` and `SP-GIST` (https://www.postgresql.org/docs/current/sql-createindex.html). I would remove these extra words: `Note: only supported by PostgreSQL`

File: activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb
Comment (fatkodima) [inline activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb:32]: I think, this should be better done in the `abstract/schema_creation.rb` base class since SQL Server also [supports](https://www.sqlservertutorial.net/sql-server-indexes/sql-server-indexes-with-included-columns/) include indexes with the same syntax.

File: activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb
Comment (fatkodima) [inline activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb:98]: Isn't existing `quoted_columns` method already covers this?

File: activerecord/test/schema/postgresql_specific_schema.rb
Comment (fatkodima) [inline activerecord/test/schema/postgresql_specific_schema.rb:145]: `supports_index_include?`?

File: activerecord/test/cases/migration/index_test.rb
Comment (fatkodima) [inline activerecord/test/cases/migration/index_test.rb:262]: `index_exists?` should be extended to also add a check for `include`. 
If you rewrite the first line with `connection.add_index("testings", "last_name")`, the second line will still pass.

**Review by fatkodima — COMMENTED** [review]
Good job 👍 Added some small comments.

File: activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb
Comment (steve-abrams) [reply activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb:98]: It is similar, but `quoted_columns` is fairly specific to the behavior surrounding the column options. It makes direct use of the `index.columns` and `index.column_options`, whereas for the `include` columns, we just need `index.include`. I considered refactoring `quoted_columns` to be a bit more generic, but the method it calls: `quoted_include_columns_for_index` also behaves slightly differently in how it handles the incoming params.

File: activerecord/test/schema/postgresql_specific_schema.rb
Comment (steve-abrams) [reply activerecord/test/schema/postgresql_specific_schema.rb]: Oof, thanks for catching this! I've fixed it.

File: activerecord/test/cases/migration/index_test.rb
Comment (steve-abrams) [reply activerecord/test/cases/migration/index_test.rb:262]: Nice catch! I updated `index_exists?` with an additional check to cover the `include` option.

File: activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb
Comment (steve-abrams) [reply activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb:32]: I should have commented this in the PR description. I saw SQL Server supports INCLUDE, but I don't have any experience with SQL Server, so my intent was to implement an initial iteration on Postgres and leave things open for SQL Server in a followup. Not sure how the maintainers like to handle things like this, but I'm a big fan of small iterations for features.

File: activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb
Comment (steve-abrams) [reply activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb]: Good point. I've updated the wording.

**Comment by steve-abrams** [comment]
Thanks for the notes @fatkodima! I responded to your comments and made a few updates based on your suggestions.

**Comment by steve-abrams** [comment]
The [build is failing](https://buildkite.com/rails/rails/builds/85785#27814880-0516-4938-af78-ebf4e1311387) and it seems unrelated to the most recent commit 🤔. I verified the failure locally, but reverting the most recent commit still failed locally with seed 20932, so I went ahead and re-pushed to see if it is a flaky failure.

File: activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb
Comment (fatkodima) [reply activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb]: You forgot to remove `UNIQUE.`

File: activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb
Comment (fatkodima) [reply activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb:32]: Ah, I see another comment from Eileen regarding this -https://github.com/rails/rails/pull/44803#discussion_r839473577.
Yes, I think this is too postgres specific, better to keep it where it is.

File: activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb
Comment (fatkodima) [reply activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb:98]: 👍 Yes, missed that it uses other index-specific settings.

**Comment by steve-abrams** [comment]
Hi @eileencodes! I just wanted to check if there was anyone I should ping to keep this moving or if we are just waiting for another reviewer to be available. No hurry, just making sure there isn't anything on my end to be done.

**Comment by fatkodima** [comment]
@steve-abrams Could you please rebase?

**Comment by fatkodima** [comment]
@yahonda Seems like the author of this PR is not very active (looking at github/twitter activity).
If there would be no activity for some time, would you mind to rebase this yourself, or I can rebase this in another PR? So this very useful feature will go into the new `7.2` version?

**Comment by steve-abrams** [comment]
Thanks for the ping @fatkodima! I'm still active, somehow I didn't get any emails or notifications until you pinged me directly. I'll see if I can update my settings so I don't miss things like this again.

@yahonda, I've rebased. The only real change as a result of the rebase was the logic I had added in `ActiveRecord::ConnectionAdapters::SchemaStatements#index_exists?` was moved to [`ActiveRecord::ConnectionAdapters::IndexDefinition#defined_for?`](https://github.com/rails/rails/pull/44803/files#diff-fac852e063eee718c2af44dec6f80b1b46f52f8aae7936e5eea6b2eab08cf2f7R52-R58). Let me know if there's anything else I can do to keep this moving.

File: activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb
Comment (yahonda) [inline activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb:767]: Would you quote `column_names` here?  I wanted every identifier to be quoted. 

`test_include_index` test generates this statement:
```sql
CREATE INDEX "include" ON "ex" ("id")INCLUDE (number)
```

- To be
```sql
CREATE INDEX "include" ON "ex" ("id")INCLUDE ("number")
```

Comment (steve-abrams) [reply activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb]: ✅

File: activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb
Comment (steve-abrams) [reply activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb]: ✅

**Comment by steve-abrams** [comment]
Thanks @yahonda, I've applied the suggested changes and rebased, would you take another look?

**Comment by yahonda** [comment]
Looks good. Would you squash your commits?

**Comment by steve-abrams** [comment]
Sure thing @yahonda, I've squashed them.

**Comment by yahonda** [comment]
Thanks!

**Pull request merged**
